### ISSUE TITLE
`invalid memory address or nil pointer dereference` panic followed by timeouts in k8s API access

### ISSUE THREAD

**Issue by ekesken** [coredns/coredns#4022]
**What happened**:

We detected the problem via `java.net.UnknownHostException` in our application side, then the observed correlated `invalid memory address or nil pointer dereference` in core-dns logs, and mitigated the issue by deleting the pods with the problem.

**What you expected to happen**:

We shouldn't have these panic errors, and '/health' endpoint should detect the failure to let self-healing.

**How to reproduce it (as minimally and precisely as possible)**:

No clue yet.

**Anything else we need to know?**:

We had API latency issues during the incident, going up to ~5s.

**Environment**:

- the version of k8s: v1.14.9-eks-f459c0
- the version of CoreDNS: v1.6.6 (image: 602401143452.dkr.ecr.eu-west-1.amazonaws.com/eks/coredns:v1.6.6)
- logs, if applicable:

```
.:53
[INFO] plugin/reload: Running configuration MD5 = 4e6d221869b89618ae45c73c146779f5
CoreDNS-1.6.6
linux/amd64, go1.12.9, 6a7a75e0
W0724 00:20:39.957007       1 reflector.go:302] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: watch of *v1.Endpoints ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 10.0.38.33:52434->172.20.0.1:443: read: connection timed out") has prevented the request from succeeding
W0724 00:20:39.957007       1 reflector.go:302] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: watch of *v1.Endpoints ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 10.0.38.33:52434->172.20.0.1:443: read: connection timed out") has prevented the request from succeeding
W0724 00:20:39.957124       1 reflector.go:302] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: watch of *v1.Service ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 10.0.38.33:52434->172.20.0.1:443: read: connection timed out") has prevented the request from succeeding
W0724 00:20:39.957124       1 reflector.go:302] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: watch of *v1.Service ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 10.0.38.33:52434->172.20.0.1:443: read: connection timed out") has prevented the request from succeeding
W0724 00:20:39.957154       1 reflector.go:302] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: watch of *v1.Namespace ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 10.0.38.33:52434->172.20.0.1:443: read: connection timed out") has prevented the request from succeeding
W0724 00:20:39.957154       1 reflector.go:302] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: watch of *v1.Namespace ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 10.0.38.33:52434->172.20.0.1:443: read: connection timed out") has prevented the request from succeeding
E0724 00:20:41.016585       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 20 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x18c4c40, 0x2dfdc70)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:51 +0x82
panic(0x18c4c40, 0x2dfdc70)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/coredns/coredns/plugin/kubernetes/object.(*Endpoints).GetNamespace(0x0, 0x0, 0x7f77a6bdc928)
	/root/src/github.com/coredns/coredns/plugin/kubernetes/object/endpoint.go:146 +0x5
k8s.io/client-go/tools/cache.MetaNamespaceKeyFunc(0x1ab18a0, 0x0, 0xc004896a40, 0x33, 0x1ab18a0, 0xc004d80b00)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/store.go:84 +0x17a
k8s.io/client-go/tools/cache.DeletionHandlingMetaNamespaceKeyFunc(0x1ab18a0, 0x0, 0x33, 0x1ab18a0, 0xc0025b1400, 0x203000)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:261 +0x6a
k8s.io/client-go/tools/cache.(*cache).Delete(0xc000585600, 0x1ab18a0, 0x0, 0x0, 0x18)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/store.go:144 +0x42
github.com/coredns/coredns/plugin/kubernetes.newdnsController.func1.1(0x18e60e0, 0xc003623140, 0x1, 0xc003623140)
	/root/src/github.com/coredns/coredns/plugin/kubernetes/controller.go:167 +0x5e7
k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc0004d5c30, 0xc0004db0e0, 0x0, 0x0, 0x0, 0x0)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/delta_fifo.go:436 +0x209
k8s.io/client-go/tools/cache.(*controller).processLoop(0xc00057bf00)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:150 +0x40
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005edfb0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:152 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0005edfb0, 0x3b9aca00, 0x0, 0xc00059e301, 0xc000088ae0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*controller).Run(0xc00057bf00, 0xc000088ae0)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:124 +0x2a8
created by github.com/coredns/coredns/plugin/kubernetes.(*dnsControl).Run
	/root/src/github.com/coredns/coredns/plugin/kubernetes/controller.go:304 +0x12b
E0724 00:20:41.016585       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 20 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x18c4c40, 0x2dfdc70)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:51 +0x82
panic(0x18c4c40, 0x2dfdc70)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/coredns/coredns/plugin/kubernetes/object.(*Endpoints).GetNamespace(0x0, 0x0, 0x7f77a6bdc928)
	/root/src/github.com/coredns/coredns/plugin/kubernetes/object/endpoint.go:146 +0x5
k8s.io/client-go/tools/cache.MetaNamespaceKeyFunc(0x1ab18a0, 0x0, 0xc004896a40, 0x33, 0x1ab18a0, 0xc004d80b00)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/store.go:84 +0x17a
k8s.io/client-go/tools/cache.DeletionHandlingMetaNamespaceKeyFunc(0x1ab18a0, 0x0, 0x33, 0x1ab18a0, 0xc0025b1400, 0x203000)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:261 +0x6a
k8s.io/client-go/tools/cache.(*cache).Delete(0xc000585600, 0x1ab18a0, 0x0, 0x0, 0x18)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/store.go:144 +0x42
github.com/coredns/coredns/plugin/kubernetes.newdnsController.func1.1(0x18e60e0, 0xc003623140, 0x1, 0xc003623140)
	/root/src/github.com/coredns/coredns/plugin/kubernetes/controller.go:167 +0x5e7
k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc0004d5c30, 0xc0004db0e0, 0x0, 0x0, 0x0, 0x0)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/delta_fifo.go:436 +0x209
k8s.io/client-go/tools/cache.(*controller).processLoop(0xc00057bf00)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:150 +0x40
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005edfb0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:152 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0005edfb0, 0x3b9aca00, 0x0, 0xc00059e301, 0xc000088ae0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*controller).Run(0xc00057bf00, 0xc000088ae0)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:124 +0x2a8
created by github.com/coredns/coredns/plugin/kubernetes.(*dnsControl).Run
	/root/src/github.com/coredns/coredns/plugin/kubernetes/controller.go:304 +0x12b
E0724 00:20:41.016585       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 20 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x18c4c40, 0x2dfdc70)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:51 +0x82
panic(0x18c4c40, 0x2dfdc70)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/coredns/coredns/plugin/kubernetes/object.(*Endpoints).GetNamespace(0x0, 0x0, 0x7f77a6bdc928)
	/root/src/github.com/coredns/coredns/plugin/kubernetes/object/endpoint.go:146 +0x5
k8s.io/client-go/tools/cache.MetaNamespaceKeyFunc(0x1ab18a0, 0x0, 0xc004896a40, 0x33, 0x1ab18a0, 0xc004d80b00)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/store.go:84 +0x17a
k8s.io/client-go/tools/cache.DeletionHandlingMetaNamespaceKeyFunc(0x1ab18a0, 0x0, 0x33, 0x1ab18a0, 0xc0025b1400, 0x203000)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:261 +0x6a
k8s.io/client-go/tools/cache.(*cache).Delete(0xc000585600, 0x1ab18a0, 0x0, 0x0, 0x18)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/store.go:144 +0x42
github.com/coredns/coredns/plugin/kubernetes.newdnsController.func1.1(0x18e60e0, 0xc003623140, 0x1, 0xc003623140)
	/root/src/github.com/coredns/coredns/plugin/kubernetes/controller.go:167 +0x5e7
k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc0004d5c30, 0xc0004db0e0, 0x0, 0x0, 0x0, 0x0)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/delta_fifo.go:436 +0x209
k8s.io/client-go/tools/cache.(*controller).processLoop(0xc00057bf00)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:150 +0x40
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005edfb0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:152 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0005edfb0, 0x3b9aca00, 0x0, 0xc00059e301, 0xc000088ae0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*controller).Run(0xc00057bf00, 0xc000088ae0)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:124 +0x2a8
created by github.com/coredns/coredns/plugin/kubernetes.(*dnsControl).Run
	/root/src/github.com/coredns/coredns/plugin/kubernetes/controller.go:304 +0x12b
E0724 00:20:41.016585       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 20 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x18c4c40, 0x2dfdc70)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:51 +0x82
panic(0x18c4c40, 0x2dfdc70)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/coredns/coredns/plugin/kubernetes/object.(*Endpoints).GetNamespace(0x0, 0x0, 0x7f77a6bdc928)
	/root/src/github.com/coredns/coredns/plugin/kubernetes/object/endpoint.go:146 +0x5
k8s.io/client-go/tools/cache.MetaNamespaceKeyFunc(0x1ab18a0, 0x0, 0xc004896a40, 0x33, 0x1ab18a0, 0xc004d80b00)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/store.go:84 +0x17a
k8s.io/client-go/tools/cache.DeletionHandlingMetaNamespaceKeyFunc(0x1ab18a0, 0x0, 0x33, 0x1ab18a0, 0xc0025b1400, 0x203000)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:261 +0x6a
k8s.io/client-go/tools/cache.(*cache).Delete(0xc000585600, 0x1ab18a0, 0x0, 0x0, 0x18)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/store.go:144 +0x42
github.com/coredns/coredns/plugin/kubernetes.newdnsController.func1.1(0x18e60e0, 0xc003623140, 0x1, 0xc003623140)
	/root/src/github.com/coredns/coredns/plugin/kubernetes/controller.go:167 +0x5e7
k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc0004d5c30, 0xc0004db0e0, 0x0, 0x0, 0x0, 0x0)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/delta_fifo.go:436 +0x209
k8s.io/client-go/tools/cache.(*controller).processLoop(0xc00057bf00)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:150 +0x40
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005edfb0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:152 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0005edfb0, 0x3b9aca00, 0x0, 0xc00059e301, 0xc000088ae0)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/root/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*controller).Run(0xc00057bf00, 0xc000088ae0)
	/root/pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/controller.go:124 +0x2a8
created by github.com/coredns/coredns/plugin/kubernetes.(*dnsControl).Run
	/root/src/github.com/coredns/coredns/plugin/kubernetes/controller.go:304 +0x12b
```

- OS (e.g: `cat /etc/os-release`):

```
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"
```

**Comment by chrisohaver** [comment]
Looks very much like the watch deletion Bugs fixed in coredns 1.7.0.

**Comment by chrisohaver** [comment]
Yes, this appears to be the tombstone deletion handling bugs in the k8s API watch that were fixed in 1.7.0.
The bug occurred after the k8s api client reconnects to k8s after a period of disconnect.

Issues: #3860 #3879
PRs #3887 #3890 #3924

**Comment by chrisohaver** [comment]
Closing. Please reopen if you find the issue still occurs with 1.7.0.
